File: src/stdpar/plugin-BeamSpotProducer/BeamSpotESProducer.cc
#include "DataFormats/BeamSpotPOD.h"
#include "Framework/ESProducer.h"
#include "Framework/EventSetup.h"
#include "Framework/ESPluginFactory.h"

#include <fstream>
#include <iostream>
#include <memory>

class BeamSpotESProducer : public edm::ESProducer {
public:
  explicit BeamSpotESProducer(std::filesystem::path const& datadir) : data_(datadir) {}
  void produce(edm::EventSetup& eventSetup);

private:
  std::filesystem::path data_;
};

void BeamSpotESProducer::produce(edm::EventSetup& eventSetup) {
  auto bs = std::make_unique<BeamSpotPOD>();

  std::ifstream in(data_ / "beamspot.bin", std::ios::binary);
  in.exceptions(std::ifstream::badbit | std::ifstream::failbit | std::ifstream::eofbit);
  in.read(reinterpret_cast<char*>(bs.get()), sizeof(BeamSpotPOD));
  eventSetup.put(std::move(bs));
}

DEFINE_FWK_EVENTSETUP_MODULE(BeamSpotESProducer);

File: src/julia-serial/plugin-BeamSpotProducer/BeamSpotESProducer.jl
using .ESPluginFactory

struct BeamSpotESProducer <: ESProducer
    data::String  # Use String to represent the path

    function BeamSpotESProducer(datadir::String)
        new(datadir)
    end
end

function readBeam(io::IOStream,es::EventSetup)

    x = read(io,Float32)
    y = read(io,Float32)
    z = read(io,Float32)
    sigmaZ = read(io,Float32)
    beamWidthX = read(io,Float32)
    beamWidthY = read(io,Float32)
    dxdz = read(io,Float32)
    dydz = read(io,Float32)
    emittanceX = read(io,Float32)
    emittanceY = read(io,Float32)
    betaStar = read(io,Float32)

    beam = BeamSpotPOD(x,y,z,sigmaZ,beamWidthX,beamWidthY,dxdz,dydz,emittanceX,emittanceY,betaStar)

    put!(es,beam)
end


function produce(producer::BeamSpotESProducer, eventSetup::EventSetup)
    beam_file = joinpath(producer.data, "beamspot.bin")
    
    #read beamspot.bin
    open(beam_file, "r") do io
        readBeam(io,eventSetup)        
    end
end


add_plugin("BeamSpotESProducer",x -> BeamSpotESProducer(x))
